Common/Orders/Slippage/MarketImpactSlippageModel.cs

```diff
@@ -0,0 +1,221 @@
+            }
+
+            // time taken for execution, we add 700ms to mimic time slippage of filling (convert to by trading day)
+            var time = ((TimeSpan)(_algorithm.UtcTime - order.CreatedTime + TimeSpan.FromMilliseconds(700))).TotalDays * 24d / 6.5d;
+            // expected valid time for impact (+ half an hour)
+            var timePost = time + 1d / 13d;
+            // normalized volume of execution
+            var nu = (double)order.AbsoluteQuantity / time / symbolData.AvgVolume;
+            // liquidity adjustment for temporary market impact, if any
+            var liquidityAdjustment = asset.Fundamentals != null ?
+                                      Math.Pow(asset.Fundamentals.CompanyProfile.SharesOutstanding / symbolData.AvgVolume, _delta) :
+                                      1d;
+            // noise adjustment factor
+            var noise = symbolData.Sigma * Math.Sqrt(timePost);
+
+            // temporary market impact
+            var permImpact = symbolData.Sigma * time * G(nu) * liquidityAdjustment + SampleGaussian(_random) * noise;
+            // permanent market impact
+            var tempImpact = symbolData.Sigma * H(nu) + SampleGaussian(_random) * noise + permImpact * 0.5d;
+            // realized market impact
+            var realizedImpact = tempImpact + permImpact;
+
+            // estimate the slippage by temporary impact
+            return SlippageFromImpactEstimation(realizedImpact);
+        }
+
+        /// <summary>
+        /// The permanent market impact function
+        /// </summary>
+        /// <param name="absOrderQuantity">The absolute, normalized order quantity</param>
+        /// <return>Unadjusted permanent market impact factor</return>
+        protected double G(double absOrderQuantity)
+        {
+            return _gamma * Math.Pow((double)absOrderQuantity, _alpha);
+        }
+
+        /// <summary>
+        /// The temporary market impact function
+        /// </summary>
+        /// <param name="absOrderQuantity">The absolute, normalized order quantity</param>
+        /// <return>Unadjusted temporary market impact factor</return>
+        protected double H(double absOrderQuantity)
+        {
+            return _eta * Math.Pow((double)absOrderQuantity, _beta);
+        }
+
+        /// <summary>
+        /// The temporary market impact function
+        /// </summary>
+        /// <param name="tempImpact">The temporary market impact</param>
+        /// <return>Slippage estimation</return>
+        protected virtual decimal SlippageFromImpactEstimation(double tempImpact)
+        {
+            // We assume an exponential distribution on the order book
+            // the order with high volume of "no slippage/impact" zone as 0, mid point of max slippage as infinity
+            // we use lambda=2.5 as parameter, so the average slippage will be (1 - 1/2.5) of the impact
+            return Convert.ToDecimal(tempImpact) * 0.6m;
```

Comment:
We should use `SafeDecimalCast`
```csharp
return tempImpact.SafeDecimalCast() * 0.6m;
```
This method should also be private or removed altogether, and `0.6` should be a class variable.